### DISCUSSION TITLE
RGB to CMYK with conversion with color map via CLI

### DISCUSSION THREAD

**Discussion by dfrumin2** [libvips/libvips#3065] [Q&A]
Hi all - just found vips and really impressed with the performance over imagemagick (which is great in itself too :-))  We are now looking to convert a series of image processing steps (CLI based) we have from IM to VIPS but not sure how to do this one step.

We are using chromium to generate a PNG file with various texts, all in primary colors (HTML is a great layout engine in this case).  This is an intermediate file that we then transform to a TIF with IM and a color map to arrive at a high resolution print file.  So think of the operation as "how can I use HTML to generate a high-res CMYKA print file".

Our command with magick looks like this:

```
magick convert ( in.png -colorspace cmyk ) \
    -write MPR:orig -alpha off  \
    -fill icc-color(cmyk,0.2,0.2,0,0) -opaque rgb(255,0,0) \
    -fill icc-color(cmyk,0,0,0,0.75) -opaque rgb(0,255,0) \
    MPR:orig -compose CopyOpacity -composite -compress zip out.tif
```

Essentially it "paints" CMYK values over each primary color (this technique is limited to 8 colors at a time) and then composites them into one image.  The result is that red in the source becomes 0.2,0.2,0,0 and green becomes 0,0,0,0.75 and then... ta-da... out.tif.

Magick is great but performance isn't that good especially with very large (30000px x 30000px = 50in square @ 600 dpi) files.

We've already seen that vips composite2 is about 5x faster than magick's composite.  I'm hoping vips can do something like the above faster than magick can.

I'm a complete noob to vips so please forgive me if this is obvious...  I was thinking this might be doable with a LUT but not sure how to put one together in the CLI.  I was also playing with draw_mask but honestly can't get that to work either, e.g.

```
vips copy in.png mask.v (this is a grayscale PNG with white text / black background)
vips copy temp.tif temp.v (this is a random TIF file for now, my idea would be to create one transparent later)
vips draw_mask temp.v "0.5 0.5 0 0.2" mask.v 0 0 
vips copy temp.v out.tif
```

I was thinking I coudl do one color at a time if I needed . But this ends up drawing white text (from mask.png) onto temp.tif without transforming the color to 0.5,0.5,0,0.2.

Would appreciate any help/pointers folks might have.  We are a .net team so we don't have folks that can do Python/Ruby but if we have to do it with the NetVips layer on top of C instead of the CLI we can probably manage.

Big thanks in advance!

**Comment by jcupitt** [comment]
Hello @dfrumin2,

I would use net-vips. The libvips CLI is pretty basic, and can't chain operations together. You have to do huge amounts of file IO, which is really painful with large images. 

In python (I'm not a net-vips expert), you could solve your problem with:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")

# remove any alpha (we only work with RGB)
if image.hasalpha():
    image = image[:-1]

# the RGB colours we test for ... you can have any number of these
rgb = [
    [255, 0, 0],     # red
    [0, 255, 0],    # green
]

# the matching CMYK densities we output, 0 - 255
cmyk = [
    [50, 50, 0, 0],    # map red to this
    [0, 0, 0, 200],    # map green to this
]

# make an array of mask images which are TRUE for pixels which equal a colour
rgb_masks = [(image == colour).bandand() for colour in rgb]

# make an array of constant CMYK images, one for each of the cmyk colours
cmyk_layers = [image.new_from_image(colour).copy(interpretation="cmyk") for colour in cmyk]

# for each pixel, find the first non-zero mask pixel, and use the corresponding cmyk pixel
output = pyvips.Image.switch(rgb_masks).case(cmyk_layers)

output.write_to_file(sys.argv[2])
```

With this test image:

![redsquare](https://user-images.githubusercontent.com/580843/192099963-f1e8205d-1092-42e4-b047-a0b4d7e9a818.svg)

I can run it like this:

```
$ ./switch.py ~/pics/redsquare.svg x.tif[compression=deflate]
```

To make:

http://www.rollthepotato.net/~john/x.tif

I don't know what your performance goals are. On this PC I see:

```
$ vips copy ~/pics/redsquare.svg[dpi=2400] x.png
$ vipsheader x.png
x.png: 19843x28063 uchar, 4 bands, srgb, pngload
$ /usr/bin/time -f %M:%e ./switch.py x.png x.tif[compression=deflate]
380984:9.72
```

So about 10s and 400mb of memory to process a 20,000 x 30,000 pixel file. It'd get a little slower with more colours.

Ask on the net-vips repo if you need help translating the code, but it should be straightforward.

**Reply by dfrumin2** [reply]
Thanks John for such a quick reply.  Two favors if you don't mind...

1 - I am an old C dev.  Any chance you can frame the code above in C?  I think I can more easily switch from C to C# than from Python.
2 - For curiosity sake, can you run your code on this 30Kx30K image?  Curious how long it takes.  I can run it through IM and even though our two PCs aren't the same, I'm going to assume they're both reasonably good high perf PCs :-)

THANKS!!!

![test copy](https://user-images.githubusercontent.com/114293525/192107219-8d76f0c8-49fb-46cd-8f5c-950155452e0b.png)

**Reply by jcupitt** [reply]
The python should be very close to C#, there will just be some minor syntax differences. C would be pretty different. 

I tried your PNG:

```
$ time ./switch.py ~/pics/test-copy.png x.tif[compression=deflate]
real	0m16.681s
user	1m28.127s
sys	0m1.074s
```

So about 17s. How long does IM take?

**Reply by jcupitt** [reply]
This PC is a 16-core threadripper workstation, though most time here is spent in PNG and TIFF load and save, so there's not a huge amount of parallelism. For example, if I don't use PNG I see:

```
$ vips copy ~/pics/test-copy.png x.tif
$ time ./switch.py x.tif y.tif 
real	0m6.941s
user	0m43.852s
sys	0m4.103s
```

Now it's 7s for your test image.

**Reply by dfrumin2** [reply]
Thanks John, I'm not sure I follow this last script with the copy .png .tif.  I'm assuming that was illustration to show that ~10s were spent in that step and 7s in the map conversion, right?  Or is there some other optimization I can make by pre-converting the PNG to a TIF?  I suspect the ICC transform from RGB to CMYK is consistent, so if it's faster I could transform the PNG to TIF first and then do a from <=> to TIF map.  thoughts?  thanks again for your absurdly fast replies :-)

**Reply by jcupitt** [reply]
Ah it's complex. Runtime is a combination of the serial and the parallel parts of the program (see [Amdahl's Law](https://en.wikipedia.org/wiki/Amdahl%27s_law)), so it's not as simple as saying "10s spent in IO". 

I was trying to show that something other than PNG (a notoriously slow file format) could be a better choice as an intermediate. Of course it depends what you can write from your HTML renderer.

**Reply by dfrumin2** [reply]
Yea, makes sense...  I don't think I have much of an option on PNG since I'm generating the source image with Chromium (to use its HTML layout engine).  So it's PNG or JPG and JPG is lossy so for small text it's not an option. So I have to chew it up...  I'm assuming any conversion of PNG => XXX will walk through the PNG bytes so no benefit unless you think the actual color mapping can be done faster in a TIF to TIF world?  Or more parallelized with threads?

BTW, on my machine (which i7-11 / 32GB mem) IM takes ~8min to do the color mapping.  So I'm *very* optimistic this will be a huge perf gain.  Still working on that C# conversion though.  jajajaj :-)

**Comment by m761372** [comment]
Hello @jcupitt, how are you?

I'm @dfrumin2 co-worker. And I am going to code an app in NetVips.

I've done tests in python with exactly proposed code.

-I trying to convert a .PNG image red/green and transparent background to a .TIF file using your python script.

But the result is not as expected.

-The background turns gray, and color green too.

Could you tell me why the transparency in the background is lost and green color turn gray?

Thanks for your help.

I attach the converted (uploaded to a server) and original file.

**Converted**: [https://easyupload.io/9m89ym](url)

**Original:**
![test](https://user-images.githubusercontent.com/66138931/195113138-4b29a873-23fc-4a28-a2c2-bc19322366ee.png)

**Reply by jcupitt** [reply]
Hi @m761372,

That python script only does RGB->CMYK -- it removes any alpha. If you want to make CMYKA, you'd need to extract the alpha at the start and reattach near the end, perhaps:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")

# remove any alpha (we only work with RGB)
alpha = None
if image.hasalpha():
    alpha = image[-1]
    image = image[:-1]

# the RGB colours we test for ... you can have any number of these
rgb = [
    [255, 0, 0],     # red
    [0, 255, 0],    # green
]

# the matching CMYK densities we output, 0 - 255
cmyk = [
    [50, 50, 0, 0],    # map red to this
    [0, 0, 0, 200],    # map green to this
]

# make an array of mask images which are TRUE for pixels which equal a colour
rgb_masks = [(image == colour).bandand() for colour in rgb]

# make an array of constant CMYK images, one for each of the cmyk colours
cmyk_layers = [image.new_from_image(colour).copy(interpretation="cmyk") for colour in cmyk]

# for each pixel, find the first non-zero mask pixel, and use the corresponding cmyk pixel
output = pyvips.Image.switch(rgb_masks).case(cmyk_layers)

# if there was alpha, reattach it
if alpha != None:
    output = output.bandjoin(alpha)

output.write_to_file(sys.argv[2])
```

The green is there in the output, it's just hard to see against the black background. Try changing the mapping for green to make it nmore visible, perhaps:

```python
cmyk = [
    [50, 50, 0, 0],    # map red to this
    [0, 255, 255, 200],    # map green to this
]
```

**Reply by m761372** [reply]
Thanks for the quick reply!

It seems to work fine with that adjustment.

**Reply by m761372** [reply]
Hello @jcupitt ! me again...

We have a problem when we convert a RGB to white (cmyk: 0,0,0,0) we need it to be visible because the sheet to be printed is colored.

It doesn't appear in the output tif. Is there a way to keep the white color?

Thanks for your help.

**Reply by jcupitt** [reply]
Do you mean you are printing on coloured paper? Won't that need a white spot colour?

**Reply by dfrumin2** [reply]
Hi John, this is a transparent TIF layer with white text that is then overlaid onto another layer that has a color background.  So we're printing on white stock but need the white text to have actual data (vs. transparent) so when we overlay on the color layer it's there.

**Reply by jcupitt** [reply]
Ah I see. Then I guess you want to test for A as well in the masks, eg.:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")

# the RGBA colours we test for ... you can have any number of these
rgba = [
    [255, 255, 255, 255], # white
    [255, 0, 0, 255],     # red
    [0, 255, 0, 255],     # green
]

# the matching CMYKA densities we output, 0 - 255
cmyka = [
    [0, 0, 0, 0, 255],      # map white to this
    [50, 50, 0, 0, 255],    # map red to this
    [0, 0, 0, 200, 255],    # map green to this
]

# make an array of mask images which are TRUE for pixels which equal a colour
rgba_masks = [(image == colour).bandand() for colour in rgba]

# make an array of constant CMYKA images, one for each of the cmyka colours
cmyka_layers = [image.new_from_image(colour).copy(interpretation="cmyk") 
                for colour in cmyka]

# for each pixel, find the first non-zero mask pixel, and use the corresponding cmyk pixel
output = pyvips.Image.switch(rgba_masks).case(cmyka_layers)

output.write_to_file(sys.argv[2])
```

(untested)

**Reply by dfrumin2** [reply]
Hi John - forgive me, we were looking in the wrong place...  The transform code that you provided earlier works great.  The issue happens when we try to join two TIFs like so:

vips composite2 5bf51b62-21cb-4ece-ab9f-c1c59eef9f74.tif 9552fefd-6a03-4527-8416-1fa5b1828ee6.tif out.tif[compression=deflate] overlay --compositing-space cmyk

For clarity, *74.tif is a transparent TIF with a graphic and *e6.tif is the transparent TIF with the text.  And the text doesn't sit on top of the graphic in *74.tif (not sure if that affects anything).

Any ideas?  I'm sure it's something silly we're doing :-)

thanks!

**Reply by jcupitt** [reply]
I'm not sure I understand the problem exactly. Could you post some sample images and the expected output?

**Reply by dfrumin2** [reply]
Thanks John - we've been working on isolating the difference.  Looks like the issue is when we try to composite2 an 8bit CMYK image with white text onto a 16bit CMYK image.  If I transform (usimg magick convert -depth 8 -compress zip) the failing base layer image to 8 bit then it seems to work.  Not sure if it's a question of size (300MB v 90MB base file) or bits...

So, follow up question - is there an equivalent vips.exe to convert 16bit to 8bit?  Then we can do that conversion before the composite2 and hopefully be good :-)

thanks!!!
